README.md
# MMM-mvgmunich [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/mrVragec/MMM-mvgmunich/raw/master/LICENSE) [![Build Status](https://api.travis-ci.org/mrVragec/MMM-mvgmunich.svg?branch=master)](https://travis-ci.org/mrVragec/MMM-mvgmunich) [![Known Vulnerabiliries](https://snyk.io/test/github/mrvragec/mmm-mvgmunich/badge.svg)](https://snyk.io/test/github/mrvragec/mmm-mvgmunich) [![Code Climate](https://codeclimate.com/github/mrVragec/MMM-mvgmunich/badges/gpa.svg)](https://codeclimate.com/github/mrVragec/MMM-mvgmunich)

MagicMirror² Module to monitor public transport (U-bahn, tram, bus, S-Bahn) in Munich - Germany.

## Example

|     |     |
| --- | --- |
|  **Stations names with arrival time**  | ![](.github/Example_1.png) | 
```
config: {
    haltestelle: "Hauptbahnhof",
    maxEntries: 10,             // 10 items on screen
    updateInterval: 60000,      // 60 s
    showIcons: false,            // Show transport type icon
    transportTypesToShow: {
      "ubahn": true,            // show ubahn route
      "sbahn": false,            // show sbahn route
      "bus": false,              // show bus route
      "regional_bus": false,     // show regional bus route
      "tram": false              // show tram route
    },
    ignoreStations: [],         // destination not to be shown
    timeToWalk: 0,             // 10 min walking time to station. Default is 0
    showWalkingTime: false,     // if the walking time should be included and the starting time is displayed
    showTrainDepartureTime: true,     // show tran departure time
    trainDepartureTimeFormat: "relative",   // format of the train departure time
    walkingTimeFormat: "relative",    // format of the walking time
  }
```
|     |     |
| --- | --- |
| **Stations names with icons and arrival time** | ![](.github/Example_3.png) |

```
config: {
  haltestelle: "Karlsplatz",
  maxEntries: 15,             // 10 items on screen
  updateInterval: 60000,      // 60 s
  showIcons: true,            // Show transport type icon
  transportTypesToShow: {
    "ubahn": true,            // show ubahn route
    "sbahn": true,            // show sbahn route
    "bus": false,              // show bus route
    "regional_bus": false,     // show regional bus route
    "tram": true              // show tram route
  },
  ignoreStations: [],         // destination not to be shown
  timeToWalk: 0,             // 10 min walking time to station. Default is 0
  showWalkingTime: true,     // if the walking time should be included and the starting time is displayed
  showTrainDepartureTime: true,     // show tran departure time
  trainDepartureTimeFormat: "relative",   // format of the train departure time
  walkingTimeFormat: "relative"    // format of the walking time
}
```
|     |     |
| --- | --- |
|  **Stations names with icons, arrival time and walking time** | ![](.github/Example_4.png) |
```
config: {
  haltestelle: "Karlsplatz",
  maxEntries: 15,             // 10 items on screen
  updateInterval: 60000,      // 60 s
  showIcons: true,            // Show transport type icon
  transportTypesToShow: {
    "ubahn": true,            // show ubahn route
    "sbahn": true,            // show sbahn route
    "bus": true,              // show bus route
    "regional_bus": false,     // show regional bus route
    "tram": true              // show tram route
  },
  ignoreStations: [],         // destination not to be shown
  timeToWalk: 0,             // 10 min walking time to station. Default is 0
  showWalkingTime: true,     // if the walking time should be included and the starting time is displayed
  showTrainDepartureTime: true,     // show tran departure time
  trainDepartureTimeFormat: "absolute",   // format of the train departure time
  walkingTimeFormat: "relative"    // format of the walking time
}
```

|     |     |
| --- | --- |
| **Stations names with icons, arrival time and disruption marking** | ![](.github/Example_5.png) |
```
config: {
  haltestelle: "Harras",
  maxEntries: 10,             // 10 items on screen
  updateInterval: 60000,      // 60 s
  showIcons: true,            // Show transport type icon
  transportTypesToShow: {
    "ubahn": true,            // show ubahn route
    "sbahn": true,            // show sbahn route
    "bus": true,              // show bus route
    "regional_bus": false,     // show regional bus route
    "tram": false              // show tram route
  },
  ignoreStations: [],         // destination not to be shown
  timeToWalk: 0,             // 10 min walking time to station. Default is 0
  showWalkingTime: false,     // if the walking time should be included and the starting time is displayed
  showTrainDepartureTime: true,     // show tran departure time
  trainDepartureTimeFormat: "absolute",   // format of the train departure time
  walkingTimeFormat: "relative",    // format of the walking time
  showInterruptions: true,
  showInterruptionsDetails: false,
  countInterruptionsAsItemShown: false
```

|     |     |
| --- | --- |
| **Stations names with icons, arrival time, disruption marking and disruptions details** (disruption details are not counted as new lines)| ![](.github/Example_6.png) |
```
config: {
  haltestelle: "Harras",
  maxEntries: 10,             // 10 items on screen
  updateInterval: 60000,      // 60 s
  showIcons: true,            // Show transport type icon
  transportTypesToShow: {
    "ubahn": true,            // show ubahn route
    "sbahn": true,            // show sbahn route
    "bus": true,              // show bus route
    "regional_bus": false,     // show regional bus route
    "tram": false              // show tram route
  },
  ignoreStations: [],         // destination not to be shown
  timeToWalk: 0,             // 10 min walking time to station. Default is 0
  showWalkingTime: false,     // if the walking time should be included and the starting time is displayed
  showTrainDepartureTime: true,     // show tran departure time
  trainDepartureTimeFormat: "absolute",   // format of the train departure time
  walkingTimeFormat: "relative",    // format of the walking time
  showInterruptions: true,
  showInterruptionsDetails: true,
  countInterruptionsAsItemShown: false
```

|     |     |
| --- | --- |
| **Stations names with icons, arrival time, disruption marking and disruptions details** (disruption details are counted as new lines) | ![](.github/Example_7.png) |
```
config: {
  haltestelle: "Harras",
  maxEntries: 10,             // 10 items on screen
  updateInterval: 60000,      // 60 s
  showIcons: true,            // Show transport type icon
  transportTypesToShow: {
    "ubahn": true,            // show ubahn route
    "sbahn": true,            // show sbahn route
    "bus": true,              // show bus route
    "regional_bus": false,     // show regional bus route
    "tram": false              // show tram route
  },
  ignoreStations: [],         // destination not to be shown
  timeToWalk: 0,             // 10 min walking time to station. Default is 0
  showWalkingTime: false,     // if the walking time should be included and the starting time is displayed
  showTrainDepartureTime: true,     // show tran departure time
  trainDepartureTimeFormat: "absolute",   // format of the train departure time
  walkingTimeFormat: "relative",    // format of the walking time
  showInterruptions: true,
  showInterruptionsDetails: true,
  countInterruptionsAsItemShown: true
```

## Dependencies
* instance of [MagicMirror²](https://github.com/MichMich/MagicMirror)
* urlencode (linux: npm install urlencode)

## Installation
1. Clone this repository in your MagicMirror installation into the folder modules.
2. Rename plugin foler from 'MMM-mvgmunich' to 'mvgmunich'
3. Install dependencies in main MagicMirror folder
4. Add configuration block to your config.js:
```
{
  module: "mvgmunich",
  position: "bottom_left",
  header: "MVG",
    config: {
      haltestelle: "Hauptbahnhof",
      maxEntries: 10,             // 10 items on screen
      updateInterval: 60000,      // 60 s
      showIcons: true,            // Show transport type icon
      transportTypesToShow: {
        "ubahn": true,            // show ubahn route
        "sbahn": true,            // show sbahn route
        "bus": true,              // show bus route
        "regional_bus": true,     // show regional bus route
        "tram": true              // show tram route
      },
      ignoreStations: [],         // destination not to be shown
      timeToWalk: 10,             // 10 min walking time to station. Default is 0
      showWalkingTime: false,     // if the walking time should be included and the starting time is displayed
      showTrainDepartureTime: true,             // show tran departure time
      trainDepartureTimeFormat: "relative",     // format of the train departure time
      walkingTimeFormat: "relative",            // format of the walking time
      showInterruptions: true,				    // show interruptions as gray-out rows
	  showInterruptionsDetails: false,		    // show details of interruptions in next line
	  countInterruptionsAsItemShown: false,	    // count interruptions details lines as a line shown
    }
},
```

## Config Options
| **Option**        | **Description** |
| --- | --- |
| `haltestelle`     | Station for which you want to display data. <br> **Default:** `Hauptbahnhof` <br> **Source:** http://www.mvg-live.de/MvgLive/MvgLive.jsp |
| `maxEntries`      | Number of items shown in table. <br> **Default:** `8` |
| `updateInterval`  | Update interval <br> **Default:** `60000` |
| `ubahn`           | Show data for U-Bahn. <br> **Possible values:** `true` or `false` <br> **Default:** `true` |
| `bus`             | Show data for Bus. <br> **Possible values:** `true` or `false` <br> **Default:** `true` |
| `regional_bus`    | Show data for regional Bus. <br> **Possible values:** `true` or `false` <br> **Default:** `true` |
| `tram`            | Show data for Tram. <br> **Possible values:** `true` or `false` <br> **Default:** `true` |
| `sbahn`           | Show data for S-Bahn. <br> **Possible values:** `true` or `false` <br> **Default:** `true` |
| `ignoreStations`  | Ignore destinations based on a array list. <br> **Possible values e.g.:** `["Feldmoching", "Hauptbahnhof"]` <br> **Default** `[]` |
| `timeToWalk`      | Time to walk to the station from your current location <br> **Default:** `0` minutes |
| `showWalkingTime`  | If the time to leave should be displayed which includes the walking time. <br> **Possible values:** `true` or `false` <br> **Default:** `false` |
| `showTrainDepartureTime` | If the time of train departure should be displayed. <br> **Possible values:** `true` or `false` <br> **Default:** `true` |
| `trainDepartureTimeFormat` | Train departure time format. Absolute: 21:10; Relative: in 8 min; <br> **Possible values:** `relative` or `absolute` <br> **Default** `relative` |
| `walkingTimeFormat` | Walking time format. Absolute: 21:08; Relative in 6 min; <br> **Possible values:** `relative` or `absolute` <br> **Default** `relative` |
| `showInterruptions` | Gray-out lines with interruptions; <br> **Possible values:** `true` or `false` <br> **Default** `false` |
| `showInterruptionsDetails` | Show interruptions details in a new line after the item with interruptions; <br> **Possible values:** `true` or `false` <br> **Default** `false` |
| `countInterruptionsAsItemShown` | Count interruption detail lines as line of a transport (number of lines will not explode in case of many interruptions; <br> **Possible values:** `true` or `false` <br> **Default** `false`  |


CHANGELOG.md
## [1.0.8] - 2019-12-04
- add option to include or exclude line based on line name.

## [1.0.7] - 2019-11-10
- fix a problem with loading plugin more then once
- fix problem with counting visible stations
- change MVG API URL
- fix some typos

## [1.0.6] - 2019-09-17
- add option to show route for regional bus

## [1.0.5] - 2019-02-20
- fix bug #20 - Stations containing Umlauts are no more recognized
- fix code climate reported issues

## [1.0.4] - 2018-11-27
- change data source URL
- add icons for transport type
- add option to show/hide transport type icons
- small refactoring

## [1.0.3] - 2017-12-15
- change approach how to read data from MVG
- optimize couple of lines
- change approach of sending data between main js and helper

## [1.0.2] - 2017-12-01
- improve approach how to handle parsed data
- add sorting of data before showing them on UI
- add option to load module more than once

## [1.0.1] - 2017-05-14
- add error handling
- add "error" handling in case of the wrong name for U-Bahn/Bus/Tram or S-Bahn station (there will be a list of all possible/similar stations)

## [1.0.0] - 2017-04-19
### Initial release
- commit source code
- add README, CHANGELOG and LICENSE files


node_helper.js
/*
 * Magic Mirror
 * Node Helper: mvgmunich
 *
 * By Simon Crnko
 * MIT Licensed
 *
 */

const NodeHelper = require("node_helper");
const request = require("request");
const urlencode = require("urlencode");
const globals = {
	"User-Agent": "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.110 Safari/537.36",
	"Content-Type": "application/x-www-form-urlencoded",
	"Accept": "application/json, text/javascript, */*; q=0.01",
	"X-Requested-With": "XMLHttpRequest",
	"X-MVG-Authorization-Key": "5af1beca494712ed38d313714d4caff6",
	"Referer": "https://www.mvg.de/dienste/abfahrtszeiten.html",
	"Accept-Encoding": "gzip",
	"Accept-Language": "en-US,en;q=0.9,de;q=0.8"
};

module.exports = NodeHelper.create({

	socketNotificationReceived: function (notification, payload) {
		// console.log("Notification in node_helper: " + notification + " - " + payload);
		const self = this;
		switch (notification) {
		case "GET_STATION_INFO":
			self.getStationInfo(payload);
			break;
		case "GET_DEPARTURE_DATA":
			self.getDepartureInfo(payload);
			break;
		default:
			console.error("Switch item {} is missing", notification);
		}
	},

	getDepartureInfo: function (payload) {
		const self = this;
		// console.log("Haltestelle: {} URL: {}", payload.haltestelle, payload.apiBase + payload.haltestelleId + "?footway=" + payload.timeToWalk);
		request({
			headers: globals,
			uri: payload.apiBase + payload.haltestelleId + "?footway=" + payload.timeToWalk,
			method: "GET",
			gzip: true
		}, function (error, response, body) {
			// console.log("Response: " + response);
			// console.log("Body: " + body);
			if (error) {
				// Error while reading departure data ...
				console.error("Error while reading departure info", error);
				self.sendSocketNotification("ERROR", "Error while reading data: " + error.message);
			} else {
				// body is the decompressed response body
				try {
					// console.log("haltestelle: {} Body: {}", payload.haltestelle, body);
					// console.log("haltestelle: {} jsonObject: {}", payload.haltestelle, jsonObject);
					// payload.transport = self.getHtml(JSON.parse(body), payload);
					payload.transport = JSON.parse(body);
					self.sendSocketNotification("UPDATE_DEPARTURE_INFO", payload);
				} catch (e) {
					console.error("Error while parsing and sending departure info", e);
					self.sendSocketNotification("ERROR_NO_DEPARTURE_DATA", "");
				}
			}
		});
	},

	getStationInfo: function (payload) {
		const self = this;
		// console.log("Station info Query: {}", payload.stationQuery + urlencode(payload.haltestelle));
		request({
			headers: globals,
			uri: payload.stationQuery + urlencode(payload.haltestelle),
			method: "GET",
			gzip: true
		}, function (error, response, body) {
			if (error) {
				// Error while reading station data ...
				console.error("Error while reading station data", error);
				self.sendSocketNotification("ERROR", "Error while reading data: " + error.message);
			} else {
				// body is the decompressed response body
				try {
					const jsonObject = JSON.parse(body);
					if (jsonObject.locations[0].id === undefined) {
						self.sendSocketNotification("ERROR_NO_STATION", "");
					} else {
						// console.log("json: {}", jsonObject.locations[0]);
						payload.haltestelleId = jsonObject.locations[0].id;
						payload.haltestelleName = jsonObject.locations[0].name;
						self.sendSocketNotification("UPDATE_STATION", payload);
					}
				} catch (e) {
					console.error("Error while parsing and sending station info", e);
					self.sendSocketNotification("ERROR_NO_STATION", "");
				}
			}
		});
	},

});


mvgmunich.js
/* Timetable for public transport in Munich */

/*
 * Magic Mirror
 * Module: MVG Munich
 *
 * By Simon Crnko
 * MIT Licensed
 *
 */

const MS_PER_MINUTE = 60000;
const mvgAPI = "https://www.mvg.de/api";
Module.register("mvgmunich", {
	// Default module configuration
	defaults: {
		maxEntries: 8, // maximum number of results shown on UI
		updateInterval: MS_PER_MINUTE, // update every 60 seconds
		apiBase: mvgAPI + "/fahrinfo/departure/",
		stationQuery: mvgAPI + "/fahrinfo/location/queryWeb?q=",
		haltestelle: "Hauptbahnhof", // default departure station
		haltestelleId: 0,
		haltestelleName: "",
		ignoreStations: [], // list of destination to be ignored in the list
		timeToWalk: 0, // walking time to the station
		showWalkingTime: false, // if the walking time should be included and the starting time is displayed
		showTrainDepartureTime: true,
		trainDepartureTimeFormat: "relative",
		walkingTimeFormat: "relative",
		showIcons: true,
		transportTypesToShow: {
			"ubahn": true,
			"sbahn": true,
			"regional_bus": true,
			"bus": true,
			"tram": true
		}
	},

	getStyles: function () {
		return ["mvgmunich.css"];
	},

	// Load translations files
	getTranslations: function () {
		return {
			en: "translations/en.json",
			de: "translations/de.json"
		};
	},

	start: function () {
		this.resultData = [];
		Log.info("Starting module: " + this.name + ", identifier: " + this.identifier);
		if (this.config.haltestelle !== "") {
			this.sendSocketNotification("GET_STATION_INFO", this.config);
		}
	},

	/*
	 * getData
	 * function call getData function in node_helper.js
	 *
	 */
	getData: function () {
		const self = this;
		self.sendSocketNotification("GET_DEPARTURE_DATA", self.config);
		setInterval(function () {
			self.sendSocketNotification("GET_DEPARTURE_DATA", self.config);
		}, self.config.updateInterval);
	},

	// Override dom generator.
	getDom: function () {
		let wrapperTable = document.createElement("div");
		if (this.config.haltestelle === "") {
			wrapperTable.className = "dimmed light small";
			wrapperTable.innerHTML = "Please set value for 'Haltestelle'.";
			return wrapperTable;
		}
		// console.log("this.resultData: {}", this.resultData);
		if (this.resultData === []) {
			wrapperTable.className = "dimmed light small";
			wrapperTable.innerHTML = "Loading data from MVG ...";
			return wrapperTable;
		}
		wrapperTable = document.createElement("table");
		wrapperTable.className = "small";
		wrapperTable.innerHTML = this.resultData[this.config.haltestelle];
		return wrapperTable;
	},

	getHtml: function (jsonObject) {
		let htmlText = "";

		// console.log("payload.maxEntries: " + payload.maxEntries);
		let visibleLines = 0;
		for (let i = 0; i < jsonObject.departures.length; i++) {
			if (visibleLines >= this.config.maxEntries) {
				break;
			}
			// get one item from api result
			const apiResultItem = jsonObject.departures[i];
			// get transport type
			const transportType = apiResultItem.product.toLocaleLowerCase();

			// console.log("transportType: " + transportType);
			// console.log("apiResultItem.destination: " + apiResultItem.destination);
			// console.log("apiResultItem.departureTime: " + apiResultItem.departureTime);

			// check if we should show data of this transport type
			// check if current station is not part of the ignore list
			if (!this.config.transportTypesToShow[transportType] ||
				this.config.ignoreStations.includes(apiResultItem.destination)) {
				continue;
			}

			htmlText += "<tr class='normal'>";
			// check if user want's icons
			htmlText += this.showIcons(apiResultItem.product, this.config.showIcons);
			// add transport number
			htmlText += "<td>" + apiResultItem.label + "</td>";
			// add last station aka direction
			htmlText += "<td class='stationColumn'>" + apiResultItem.destination + "</td>";
			// check if user want's to see departure time
			htmlText += this.showDepartureTime(apiResultItem.departureTime);
			// check if user want's to see walking time
			htmlText += this.showWalkingTime(apiResultItem.departureTime);
			htmlText += "</tr>";
			visibleLines++;
		}
		// console.log("htmlText: " + "haltestelle: " + payload.haltestelle + " - " + htmlText);
		return htmlText;
	},

	showIcons: function (product, showIcons) {
		// console.log("Show icons: ", showIcons);
		let icons = "";
		if (showIcons) {
			icons = "<td class='" + product.toLocaleLowerCase() + "'></td>";
		}
		// console.log("Icons content: {}", icons);
		return icons;
	},

	showWalkingTime: function (departureTime) {
		let htmlText = "";
		if (this.config.showWalkingTime) {
			htmlText += "<td> / ";
			const startWalkingTime = new Date(departureTime - this.config.timeToWalk * MS_PER_MINUTE);
			// check what kind of walking time user wants (absolute / relative)
			if (this.config.walkingTimeFormat === "absolute") {
				htmlText += this.getAbsoluteTime(startWalkingTime);
			} else if (this.config.walkingTimeFormat === "relative") {
				htmlText += this.getRelativeTime(startWalkingTime);
			} else {
				htmlText += "walkingTimeFormat config is wrong";
			}
			htmlText += "</td>";
		}
		return htmlText;
	},

	showDepartureTime: function (departureTime) {
		let htmlText = "";
		if (this.config.showTrainDepartureTime) {
			// add departure time
			htmlText += "<td class='timing'>";
			const departureDate = new Date(departureTime);
			// console.log("departureDate: " + departureDate);
			// check what kind of time user wants (absolute / relative)
			if (this.config.trainDepartureTimeFormat === "absolute") {
				htmlText += this.getAbsoluteTime(departureDate);
			} else if (this.config.trainDepartureTimeFormat === "relative") {
				htmlText += this.getRelativeTime(departureDate);
			} else {
				htmlText += "trainDepartureTimeFormat config is wrong";
			}
			htmlText += "</td>";
		}
		return htmlText;
	},

	getAbsoluteTime: function (time) {
		let hoursStr = (time.getHours() < 10 ? "0" : "") + time.getHours();
		let minutesStr = (time.getMinutes() < 10 ? "0" : "") + time.getMinutes();

		return hoursStr + ":" + minutesStr;
	},

	getRelativeTime: function (time) {
		const timingForStartWalking = Math.floor((time.getTime() - new Date().getTime()) / 1000 / 60);
		return (timingForStartWalking <= 0
			? this.translate("JETZT")
			: this.translate("IN") + " " + timingForStartWalking + " " + this.translate("MIN"));
	},

	// Override getHeader method.
	getHeader: function () {
		if (this.config.haltestelle !== "" || this.config.haltestelleName !== "") {
			return this.data.header + " Munich: " +
				(this.config.haltestelleName === "" ? this.config.haltestelle : this.config.haltestelleName);
		}
		return "";
	},

	socketNotificationReceived: function (notification, payload) {
		// console.log("Notification in mvgmunich: " + notification + " - " + payload);
		// console.log("this.config.haltestelle: " + this.config.haltestelle);
		// console.log("payload.transport: " + payload.transport);
		switch (notification) {
		case "UPDATE_DEPARTURE_INFO":
			this.resultData[payload.haltestelle] = this.getHtml(payload.transport);
			break;
		case "UPDATE_STATION":
			if (this.config.haltestelle === payload.haltestelle) {
				this.config.haltestelleId = payload.haltestelleId;
				this.config.haltestelleName = payload.haltestelleName;
			}
			this.getHeader();
			this.getData();
			break;
		default:
			Log.error();
		}
		this.updateDom();
	}
});
